scio-core/src/main/scala/com/spotify/scio/values/PairHashSCollectionFunctions.scala

```diff
@@ -37,23 +37,53 @@ class PairHashSCollectionFunctions[K, V](val self: SCollection[(K, V)]) {
   def hashJoin[W: Coder](
     that: SCollection[(K, W)]
   )(implicit koder: Coder[K], voder: Coder[V]): SCollection[(K, (V, W))] =
-    hashJoin(SideMap(combineAsMapSideInput(that)))
+    hashJoin(that.asMultiMapSideInput)
 
   /**
-   * Perform an inner join with a SideMap.
+   * Perform an inner join with a [[SideMap]].
+   *
+   * `SideMap`s are deprecated in favor of `SideInput[Map[K, Iterable[W]]]`.
+   * Example replacement:
+   * {{{
+   *   val si = pairSCollRight.asMultiMapSideInput
+   *   val joined1 = pairSColl1Left.hashJoin(si)
+   *   val joined2 = pairSColl2Left.hashJoin(si)
+   * }}}
    *
    * @group join
    */
+  @deprecated(
+    "Use SCollection[(K, V)]#hashJoin(that) or SCollection[(K, V)]#hashJoin(that.asMultiMapSideInput) instead.",
+    "0.8.0"
+  )
   def hashJoin[W: Coder](
-    that: SideMap[K, W]
+    sideMap: SideMap[K, W]
+  )(implicit koder: Coder[K], voder: Coder[V]): SCollection[(K, (V, W))] =
+    hashJoin(sideMap.asImmutableSideInput)
+
+  /**
+   * Perform an inner join with a MultiMap `SideInput[Map[K, Iterable[V]]`
+   *
+   * The right side is tiny and fits in memory. The SideInput can be used reused for
+   * multiple joins.
+   *
+   * Example:
+   * {{{
+   *   val si = pairSCollRight.asMultiMapSideInput
+   *   val joined1 = pairSColl1Left.hashJoin(si)
+   *   val joined2 = pairSColl2Left.hashJoin(si)
+   * }}}
+   *
+   * @group join
+   */
+  def hashJoin[W: Coder](
+    thatSideInput: SideInput[Map[K, Iterable[W]]]
```

Comment:
@anish749 sorry for the nit's ... I really wanted to avoid the `that*` arg naming.

Reply:
No worries, i've fixed it now.